File: HOT100/14.贪心算法/763.划分字母区间.cpp
/*
 * @lc app=leetcode.cn id=763 lang=cpp
 *
 * [763] 划分字母区间
 */
#include "tools.h"

// @lc code=start
class Solution
{
public:
    vector<int> partitionLabels(string s)
    {
        unordered_map<char, int> hash;
        int n = s.size();
        for (int i = 0; i < n; i++)
        {
            hash[s[i]] = i;
        }
        int end = 0, start = 0;
        vector<int> res;
        for (int i = 0; i < n; i++)
        {
            end = max(end, hash[s[i]]);
            if (i == end)
            {
                res.push_back(end - start + 1);
                start = end + 1;
            }
        }
        return res;
    }
};
// @lc code=end

File: CodeTop/字节/763.划分字母区间.py
#
# @lc app=leetcode.cn id=763 lang=python3
#
# [763] 划分字母区间
#
from mytools import *
# @lc code=start
class Solution:
    def partitionLabels(self, s: str) -> List[int]:
        hash = defaultdict(int)
        for i, ch in enumerate(s):
            hash[ch] = i
        n = len(s); i = 0
        res = []
        while i < n:
            eds = hash[s[i]]
            j = i + 1
            while j <= eds:
                eds = max(eds, hash[s[j]])
                j += 1
            res.append(j - i)
            i = j
        return res
# @lc code=end

